src/cipher.rs

```diff
@@ -472,319 +486,651 @@ impl rustls_root_cert_store {
+    /// existing hint subjects. Calling this function with empty `subjects` will have no
+    /// effect, use `rustls_web_pki_client_cert_verifier_clear_root_hint_subjects` to clear
+    /// the subject hints.
+    #[no_mangle]
+    pub extern "C" fn rustls_web_pki_client_cert_verifier_add_root_hint_subjects(
+        builder: *mut rustls_web_pki_client_cert_verifier_builder,
+        store: *const rustls_root_cert_store,
+    ) -> rustls_result {
+        let client_verifier_builder: &mut Option<ClientCertVerifierBuilder> =
+            try_mut_from_ptr!(builder);
+        let client_verifier_builder = match client_verifier_builder {
+            None => return AlreadyUsed,
+            Some(v) => v,
+        };
+
+        let store = try_arc_from_ptr!(store);
+        client_verifier_builder.root_hint_subjects = store.subjects();
+        rustls_result::Ok
+    }
+
+    /// Create a new client certificate verifier from the builder.
     ///
     /// The builder is consumed and cannot be used again, but must still be freed.
     ///
-    /// The verifier can be used in several `rustls_server_config` instances. Must be freed by
-    /// the application when no longer needed. See the documentation of
-    /// `rustls_allow_any_authenticated_client_verifier_free` for details about lifetime.
-    /// This copies the contents of the `rustls_root_cert_store`. It does not take
-    /// ownership of the pointed-to memory.
+    /// The verifier can be used in several `rustls_server_config` instances and must be
+    /// freed by the application when no longer needed. See the documentation of
+    /// `rustls_web_pki_client_cert_verifier_builder_free` for details about lifetime.
     #[no_mangle]
-    pub extern "C" fn rustls_allow_any_authenticated_client_verifier_new(
-        builder: *mut rustls_allow_any_authenticated_client_builder,
-    ) -> *const rustls_allow_any_authenticated_client_verifier {
+    pub extern "C" fn rustls_web_pki_client_cert_verifier_builder_build(
+        builder: *mut rustls_web_pki_client_cert_verifier_builder,
+        verifier_out: *mut *mut rustls_client_cert_verifier,
+    ) -> rustls_result {
         ffi_panic_boundary! {
-            let client_cert_verifier_builder: &mut Option<AllowAnyAuthenticatedClient> = try_mut_from_ptr!(builder);
+            let client_verifier_builder: &mut Option<ClientCertVerifierBuilder> = try_mut_from_ptr!(builder);
+            let client_verifier_builder = try_take!(client_verifier_builder);
 
-            let client_cert_verifier = match client_cert_verifier_builder.take() {
-                None => {
-                    return null() as *const _;
-                },
-                Some(x) => x,
+            let mut builder = WebPkiClientVerifier::builder(client_verifier_builder.roots)
+                .with_crls(client_verifier_builder.crls);
+            if matches!(client_verifier_builder.revocation_depth, RevocationCheckDepth::EndEntity) {
```

Comment:
I think this would be better expressed as an exhaustive match with `RevocationCheckDepth::Chain => {}` as the other branch. That makes it clear when reading this that there's only one other possibility. Same below with UnknownStatusPolicy::Allow / Deny.

Reply:
Makes sense. I've updated each instance, for the verifier depth and unknown revocation status, for both the client cert and server cert verifier impls.